compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -9509,7 +9513,38 @@ OMR::Z::TreeEvaluator::loadaddrEvaluator(TR::Node * node, TR::CodeGenerator * cg
             if (sym && sym->isStartPC())
                {
                cursor = generateRILInstruction(cg, TR::InstOpCode::LARL, node, targetRegister, node->getSymbolReference(),
-                                                reinterpret_cast<void*>(sym->getStaticAddress()),  NULL);
+                                               reinterpret_cast<void*>(sym->getStaticAddress()),  NULL);
+               }
+            else if (cg->needRelocationsForBodyInfoData() && sym && sym->isRecompilationCounter())
+               {
+               cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
+                                                     (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
+                                                     TR_BodyInfoAddress, NULL, NULL, NULL);
+               }
+            else if (cg->needClassAndMethodPointerRelocations() && node->getSymbol()->isClassObject())
+               {
+               cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
+                                                     (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
+                                                     TR_ClassAddress, NULL, NULL, NULL);
+               }
+            else if (cg->needRelocationsForStatics() && sym)
+               {
+               int32_t reloType = 0;
+               if (sym->isCompiledMethod())
+                  {
+                  reloType = TR_RamMethod;
+                  }
+               else if (sym->isStatic() && !sym->isNotDataAddress())
+                  {
+                  reloType = TR_DataAddress;
+                  }
+
+               if (reloType != 0)
+                  {
+                  cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
+                                                        (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
+                                                        reloType, NULL, NULL, NULL);
+                  }
```

Comment:
What if `reloType == 0`? Is it an error?

Reply:
Can't recall why I added this initially. But I think we can simplify this area. There's something similar on 9559 where we set `reloType = 0` and then create a relocation record with it eventually. The problem is that `reloType=0` corresponds to `TR_ConstantPool`. So this can lead to incorrect relocations/errors.

I think it may be simpler to pull these checks out into the outer `if` tree for the `needRelocationsForStatics` and `compileRelocatableCode` blocks. Then in the fall back path we will always take the path on line 9567 below.

It would look something like this:
```
else if (cg->needClassAndMethodPointerRelocations() && sym && node->getSymbol()->isClassObject())
                {
                cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
                                                      (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
                                                      TR_ClassAddress, NULL, NULL, NULL);
                }
else if (cg->needRelocationsForStatics() && sym && sym->isCompiledMethod())
              {
              cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
                                                         (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
                                                         TR_RamMethod, NULL, NULL, NULL);
              }
else if (cg->needClassAndMethodPointerRelocations() && sym && sym->isStatic() && !sym->isNotDataAddress())
              {
              cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
                                                         (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
                                                         TR_DataAddress, NULL, NULL, NULL);
              }
else if (comp->compileRelocatableCode() && node->getSymbol()->isDebugCounter())
             {
             cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
                                                     (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
                                                     TR_DebugCounter, NULL, NULL, NULL);
             }
....
....
....
....
else
             {
               cursor = genLoadAddressConstant(cg, node, (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(), targetRegister);
             }

If you're okay with it then I'll go ahead and implement the change.

Reply:
Sure that is fine. I dislike that we are using an existing enum value as the initial value for `reloType`. Can we initialize these to some non-existent relo type (will -1 work?). Also if we have an if/else chain where at the end we are to determine a valid value for `reloType` then I'd like to see a fatal assert at the end that it is not set to the default initialized value. This will prevent future bugs in this area.

Reply:
If we're going with the above approach, then we will no longer have a `reloType` variable to worry about. But I took a look around to see how this can be handled, and this looks like a good example of how to handle similar situations:
https://github.com/eclipse/omr/blob/24cd6ede529b51b68f8af33474c7006525cb8c86/compiler/z/codegen/OMRTreeEvaluator.cpp#L264-L275

Reply:
Fix available here: b7d7a15d0dfcda95b43af2cc2ee9070212f259d4. Please take a look.